### ISSUE TITLE
Networked Fluid Transposer

### ISSUE THREAD

**Issue by ruifung** [ExtraCells/ExtraCells2#301]
1. Remove the current implementation of a Fluid Crafter
2. Add a networked version of a "Fluid Transposer" that can provide filled containers based on available fluids, and drain containers straight to storage.

Feedback is most appreciated.

The main argument against the Fluid Crafter is that it simply strips out the fluid containers and would probably stall the AE crafting with no indication when fluids are not sufficient.

**Comment by XFactHD** [comment]
This looks like you want to add an AE version of the TE fluid transposer. Very useful idea.

**Comment by ruifung** [comment]
And remove the fluid crafter since actually having a fully functional fluid crafting for AE is not worth the effort - I had a discussion with the AE maintainers on IRC yesterday. This is the best alternative I could come up with that would allow for materials calculation to calculate the fluid requirements too.

Proposed functionality:
- For example you have 3500 mB of lava in the network, the "Network Transposer", for lack of a better name - has been configured to provide buckets of lava - therefore it reports that it can provide 3 buckets of lava.
- If there are buckets in the network, it will use them otherwise it will request the network to craft if possible.
- An second mode (or another machine, maybe) will let it act as a item storage sink with configurable priority - it will accept only fluid containers, and drain those containers into fluid storage if there is space. And return the empty or partially drained containers to regular item storage. Also will accept filled containers piped directly in,

**Comment by XFactHD** [comment]
I think the name "Networked Fluid Transposer" isn't bad at all.
Also, the proposed functionality makes complete sense to me, you did a good job on that.

**Comment by Heliomance** [comment]
The ability to request filled buckets of stuff and have them be filled from storage is valuable in and of itself - I fully support this idea. One possible idea for the fluid crafting though - would it be possible to create "virtual bucket" items that are created for free by the system, filled, used for crafting, then immediately destroyed?

**Comment by ruifung** [comment]
That idea has crossed my mind, but it wouldn't be possible - I don't believe AE will let me do that? Will it? @yueh @thatsIch

**Comment by thatsIch** [comment]
maybe but is this really meaningful. If you add a machine which fills up liquids with containers, then the containers will be refunded upon crafting anyways. Else there could be duping if not done right.

I think a far more interesting part of fluid crafting would be support for TiC alloy making or stuff like **ingot cast pattern** where if you combine fluid A with a specific cast you can get the results like in a forge thingy (whatever name it was)

tl;dr; imo better keep it vanilla, saver

**Comment by Macrofarad** [comment]
Seeing as people were invited to comment on this thread again, here's my two cents:

As far as I can tell, , mods like Extra cells piggyback on AE2's backbone, but for the most part, they are actually functioning as separate systems that just happen to share channels/power/hardware.  This is where a lot of the compatibility issues likely end up coming from.  What probably needs to happen is a fluid compatibility layer.  This layer of code would essentially function as variable casting, giving the mainline AE2 system phantom items each referencing a milibucket of fluid in the Extra cells storage.  Needless to say, the player should not be able to extract these phantom items and in an ideal case, shouldn't even be able to see them.  Doing this would provide the AE2 system with something it knows how to work with.  Things such as do I have enough <insert fluid here> to craft ItemX would become much, much easier to do. 

Implementation: modifying AE2 directly is likely a no go, but it should be relatively simple to make a block that the AE2 system sees as a storage unit (Name it Fluid compatibility Layer?).  Once this block is connected it will then provide levels to the AE2 system via the phantom items.  Because these phantom items would only be references, as the fluid level goes up or down, the phantom item would also respond accordingly (and visa versa), moreover, because this block would be telling the AE2 system it was an external storage, these items would not take up any extra space.  Have them function like controllers, only one per system to prevent chaos.

Anyways, for what it's worth, that's my two cents.  Whether or not it happens or is even viable depends on the code itself.

**Comment by ruifung** [comment]
I've considered something like that, but there's pretty much no way to do phantom items, - at least not that I can think of. The main reason is that last I checked, to AE, every Item in the grid is a valid item for extraction. Something like this needs to be in AE, since the item handling code is entirely up to AE.

**Comment by ruifung** [comment]
Something like this was considered, in a way. But the closest I can come do is using a provider to provide buckets of fluids - I can't do a phantom item - AE doesn't support something like that.

**Comment by XFactHD** [comment]
@Macrofarad The "AEFluidStack" which is used to store fluids in the ME system is already a compatibility layer built into AE2  itself. The alternative of providing buckets also doesn't really work because you would either need to have empty buckets in your system at all time or you destroy the bucket item after the crafting operation, which is not really possible.

**Comment by ruifung** [comment]
, the latter actually - it acts as a autocrafting for fluid filled buckets. - But thats the closest i can come without native support for AEFluidStacks in the crafting subsystem.

**Comment by Heliomance** [comment]
Does this thing only fill buckets, or does it act as a full replacement for the Thermal Dynamics Fluid Transposer?

**Comment by XFactHD** [comment]
It does only fill buckets.

**Comment by thatsIch** [comment]
We do not own such thing as Fluid Crafting internally, because there exists no fluid crafting. EC2 implements that, so it is their decision how to distribute that. What they can do is invoke item crafting into their crafting tree as a sub tree, basically wrapping everything. 

The crafting is based on the recipe so if a bucket is a byproduct, you could technically subclass that recipe via reflection and delete the result or check via simulation if you get the same byproduct as given in the recipe. I think there are multiple routes you can take.

It is nothing easy though, because crafting in general is quite a complicated matter.
